### PR TITLE
HC-41: Add Chats Page

### PR DESCRIPTION
**Pull request by petrynchyn**

### CODE REVIEW

File: backend/src/api/api.ts [hunk 1]
```diff
@@ -9,6 +9,7 @@ import { initDiagnosisApi } from './diagnosis/diagnosis.api';
 import { initUploadFileApi } from './upload-file/upload-file.api';
 import { initPermissionApi } from './permission/permission.api';
 import { initNotificationApi } from './notification/notification.api';
+import { initChatApi } from './chat/chat.api';
```
Comment (ghost) [inline backend/src/api/api.ts]:
```suggestion
import { initMessagesApi } from './chat/chat.api';
```

File: backend/src/api/chat/chat.api.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+import { Router } from 'express';
+import { validateSchema } from '~/middlewares';
+import { message as validationMessageSchema } from '~/validation-schemas';
+import { ApiPath, HttpCode, ChatsApiPath } from '~/common/enums';
+import { chat as chatService } from '~/services/services';
+
+const initChatApi = (apiRouter: Router): Router => {
```
Comment (ghost) [inline backend/src/api/chat/chat.api.ts]:
```suggestion
const initMessagesApi = (apiRouter: Router): Router => {
```
Remove `chat` word from everywhere

File: backend/src/data/repositories/chat-repository.ts [hunk 3]
```diff
@@ -0,0 +1,125 @@
+import { Sequelize, Op } from 'sequelize';
+import { MemberModel, MessageModel } from '../models';
+import { IMember, IMessage } from '~/common/interfaces';
+
+class ChatRepository {
```
Comment (ghost) [inline backend/src/data/repositories/chat-repository.ts]:
```suggestion
class MessagesRepository {
```

File: shared/src/common/enums/api/chats-api-path.enum.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+enum ChatsApiPath {
```
Comment (ghost) [inline shared/src/common/enums/api/chats-api-path.enum.ts]:
```suggestion
enum MessagesApiPath {
```

File: frontend/src/components/chats/components/select-async/select-async.tsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
+import * as React from 'react';
```
Comment (ghost) [inline frontend/src/components/chats/components/select-async/select-async.tsx]: Remove this component

File: frontend/src/components/chats/components/message-list/styles.module.scss [hunk 6]
```diff
@@ -0,0 +1,45 @@
+.messageList {
+
+  // display: flex;
+  // flex-flow: column nowrap;
+
+  // & > :first-child {
+  //   margin-top: auto !important;
+  //   /* use !important to prevent breakage from child margin settings */
+  // }
```
Comment (ghost) [inline frontend/src/components/chats/components/message-list/styles.module.scss]: Remove comment

File: frontend/src/components/chats/components/message-list/message-list.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+            {curFromNow !== nextFromNow && true &&<HorizontalLine label={curFromNow} />}
+            <Message
+              message={text}
+              time={moment.utc(createdAt).local().format('LT')}
+              avatar={memberId === userId ? userAvatarPath : memberAvatarPath}
+              isOutcoming={memberId === userId}
+            />
+          </div>
+        );
+      }).reverse()}
```
Comment (ghost) [inline frontend/src/components/chats/components/message-list/message-list.tsx]: Don't user `reverse`
If you need message in specific order - fetch in this order

File: frontend/package.json [hunk 8]
```diff
@@ -20,6 +20,7 @@
     "leaflet": "1.7.1",
     "leaflet-lasso": "2.2.4",
     "query-string": "7.0.0",
+    "moment": "2.29.1",
```
Comment (ghost) [inline frontend/package.json]: we don't use this library

File: frontend/src/components/chats/components/message-list/message-list.tsx [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import * as React from 'react';
+import { useSelector } from 'react-redux';
+import moment from 'moment';
```
Comment (ghost) [inline frontend/src/components/chats/components/message-list/message-list.tsx]: Remove moment at all

File: frontend/src/components/chats/components/member/member.tsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+  info?: string;
+  avatar: string;
+  isSelected?: boolean;
+  className?: string;
+  onClick?: (id: string) => void;
+}
+
+const Member: React.FC<Props> = ({ id, label, info, avatar, isSelected, className, onClick }) => {
+  const handlerOnClick = () => {
+    !isSelected && onClick && onClick(id);
```
Comment (ghost) [inline frontend/src/components/chats/components/member/member.tsx]:
```suggestion
    !isSelected && onClick?.(id);
```

File: frontend/src/components/chats/components/input-message-form/input-message-form.tsx [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+  }));
+
+  const DEFAULT_VALUES: IMessagePayload = {
+    to: selectedMemberId,
+    text: '',
+  };
+  const { handleSubmit, errors, control, register, reset, formState: { isDirty } } = useForm<IMessagePayload>({
+    resolver: yupResolver(messageValidationSchema),
+    defaultValues: DEFAULT_VALUES,
+    mode: 'onChange',
```
Comment (ghost) [inline frontend/src/components/chats/components/input-message-form/input-message-form.tsx]:
```suggestion
  const DEFAULT_VALUES: IMessagePayload = {
    to: selectedMemberId,
    text: '',
  };
  const { handleSubmit, errors, control, register, reset, formState: { isDirty } } = useForm<IMessagePayload>({
    resolver: yupResolver(messageValidationSchema),
    defaultValues: {
      to: selectedMemberId,
      text: '',
    },
    mode: 'onChange',
```

File: frontend/src/components/chats/components/add-member-form/add-member-form.tsx [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+  const { handleSubmit, errors, control, reset, formState: { isDirty } } = useForm<IMemberPayload>({
+    // resolver: yupResolver(messageValidationSchema),
+    defaultValues: DEFAULT_VALUES,
+    mode: 'onChange',
+  });
+
+  const dispatch = useDispatch();
+
+  const loadOptions = (inputValue: string, callback: any) => {
```
Comment (ghost) [inline frontend/src/components/chats/components/add-member-form/add-member-form.tsx]: Remove any

File: frontend/src/components/chats/components/add-member-form/add-member-form.tsx [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+
+interface Props {
+  className?: string;
+}
+
+const AddMemberForm: React.FC<Props> = ({ className }) => {
+
+  const DEFAULT_VALUES: IMemberPayload = {
+    [MemberPayloadKey.MEMBER]: null,
+  };
```
Comment (ghost) [inline frontend/src/components/chats/components/add-member-form/add-member-form.tsx]: Move default values to `frontend/src/components/chats/components/add-member-form/common/constants/`

File: frontend/src/components/chats/components/add-member-form/add-member-form.tsx [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+import styles from './styles.module.scss';
+
+enum MemberPayloadKey {
+  MEMBER = 'member',
+}
+
+interface IMemberPayload {
+  [MemberPayloadKey.MEMBER]: IMember | null;
+}
+
```
Comment (ghost) [inline frontend/src/components/chats/components/add-member-form/add-member-form.tsx]: Clean up the component

File: frontend/src/common/interfaces/form/option.interface.ts [hunk 15]
```diff
@@ -1,6 +1,6 @@
-interface IOption<T = string> {
+interface IOption<T = string, V=any> {
```
Comment (ghost) [inline frontend/src/common/interfaces/form/option.interface.ts]: Remove it

File: backend/src/services/chat/chat.service.ts [hunk 16]
```diff
@@ -0,0 +1,21 @@
+import { chatRepository } from '~/data/repositories';
+import { IMember, IMessage } from '~/common/interfaces';
+
+class Chat {
```
Comment (ghost) [inline backend/src/services/chat/chat.service.ts]:
```suggestion
class Messages {
```

File: backend/src/data/repositories/chat-repository.ts [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+          {
+            [Op.and]: [
+              { to: userId },
+              { userId: memberId },
+            ],
+          },
+        ],
+      },
+      order: [
+        ['createdAt', 'ASC'],
```
Comment (ghost) [inline backend/src/data/repositories/chat-repository.ts]: Don't use just strings. Use enums

**Review by ghost — CHANGES_REQUESTED** [review]
Remove member model at all

File: frontend/src/components/chats/components/select-async/select-async.tsx (hunk 5 above)
Comment (petrynchyn) [reply frontend/src/components/chats/components/select-async/select-async.tsx]: I used it `const SelectAsync: React.FC<Props> = ({`

File: backend/src/api/messages/messages.api.ts [hunk 18]
```diff
@@ -0,0 +1,51 @@
...
+  apiRouter.use(ApiPath.CHATS, chatRouter);
+
+  // chatRouter.get(MessagesApiPath.MEMBERS, async (req, res, next) => {
+  //   try {
+  //     const result = await messagesService.getMembersAsChats(); //  req.user?.id ?? ''
+  //     res.status(HttpCode.OK).json(result);
+  //   } catch (error) {
+  //     next(error);
+  //   }
+  // });
```
Comment (ghost) [inline backend/src/api/messages/messages.api.ts]:
```suggestion

```

File: backend/src/api/messages/messages.api.ts [hunk 19]
```diff
@@ -0,0 +1,51 @@
+import { Router } from 'express';
+import { validateSchema } from '~/middlewares';
+import { message as validationMessageSchema } from '~/validation-schemas';
+import { ApiPath, HttpCode, MessagesApiPath } from '~/common/enums';
+import { messages as messagesService } from '~/services/services';
+
+const initMessagesApi = (apiRouter: Router): Router => {
+  const chatRouter = Router();
```
Comment (ghost) [inline backend/src/api/messages/messages.api.ts]:
```suggestion
  const messagesRouter = Router();
```
Rename all `chat` word

File: backend/src/common/enums/db/model-name.enum.ts [hunk 20]
```diff
@@ -14,6 +14,7 @@ enum ModelName {
   SPECIALIZATION = 'Specialization',
   USER_SPECIALIZATION = 'UserSpecialization',
   PROFESSION = 'Profession',
+  MEMBER = 'MemberModel',
```
Comment (ghost) [inline backend/src/common/enums/db/model-name.enum.ts]:
```suggestion
```

File: backend/src/common/enums/db/order-key.enum.ts [hunk 21]
```diff
@@ -0,0 +1,6 @@
+enum OrderKey {
```
Comment (ghost) [inline backend/src/common/enums/db/order-key.enum.ts]: We already have it as `SortType` enum so remove it

File: shared/src/common/interfaces/messages/member.instance.ts [hunk 22]
```diff
@@ -0,0 +1,9 @@
+import { MemberKey } from '~/common/enums';
+
+interface IMember {
```
Comment (ghost) [inline shared/src/common/interfaces/messages/member.instance.ts]: Remove

File: shared/src/common/enums/messages/member-key.enum.ts [hunk 23]
```diff
@@ -0,0 +1,7 @@
+enum MemberKey {
```
Comment (ghost) [inline shared/src/common/enums/messages/member-key.enum.ts]: Remove

File: shared/src/common/enums/api/messages-api-path.enum.ts [hunk 24]
```diff
@@ -0,0 +1,9 @@
+enum MessagesApiPath {
+  ROOT = '/',
+  MESSAGES = '/messages',
+  MESSAGES_$ID = '/messages/:id',
+  MEMBERS = '/members',
+  MEMBERS_$NAME = '/members/:name',
```
Comment (ghost) [inline shared/src/common/enums/api/messages-api-path.enum.ts]:
```suggestion
  USERS = '/users',
```

File: shared/src/common/enums/api/api-path.enum.ts [hunk 25]
```diff
@@ -8,6 +8,7 @@ enum ApiPath {
   DIAGNOSES = '/diagnoses',
   FILES = '/files',
   PERMISSIONS = '/permissions',
+  CHATS = '/chats',
```
Comment (ghost) [inline shared/src/common/enums/api/api-path.enum.ts]:
```suggestion
  MESSAGES = '/messages',
```

File: frontend/src/store/slices/messages/messages.slice.ts [hunk 26]
```diff
@@ -0,0 +1,122 @@
...
+//     throw error;
+//   }
+// };
+
+const loadFilteredMembersAsChats = (name: string): AppThunk => async dispatch => {
+  try {
+    const response = await messagesApi.getMembersByName(name);
+    dispatch(actions.setMembers(response));
+    response?.length
+      ? dispatch(selectMember(response[0].id))
```
Comment (ghost) [inline frontend/src/store/slices/messages/messages.slice.ts]:
```suggestion
      ? dispatch(selectMember(response.id))
```

File: frontend/src/store/slices/messages/messages.slice.ts [hunk 27]
```diff
@@ -0,0 +1,122 @@
...
+
+  } catch (error) {
+    if (error instanceof HttpError) {
+      notificationService.error(`Error ${error.status}`, error.messages);
+    }
+    throw error;
+  }
+};
+
+// const loadFilteredMembersAsChats = (name: string, callback: any): AppThunk => async () => {
```
Comment (ghost) [inline frontend/src/store/slices/messages/messages.slice.ts]: Remove

File: frontend/src/store/slices/messages/messages.slice.ts [hunk 28]
```diff
@@ -0,0 +1,122 @@
...
+
+  } catch (error) {
+    if (error instanceof HttpError) {
+      notificationService.error(`Error ${error.status}`, error.messages);
+    }
+    throw error;
+  }
+};
+
+const addMember = (member: IMember): AppThunk => dispatch => {
```
Comment (ghost) [inline frontend/src/store/slices/messages/messages.slice.ts]: Remove all member names

File: frontend/src/services/messages/messages.service.ts [hunk 29]
```diff
@@ -0,0 +1,60 @@
+import { IMember, IMessage, IMessagePayload } from 'common/interfaces';
```
Comment (ghost) [inline frontend/src/services/messages/messages.service.ts]: Remove all member names

File: frontend/src/components/messages/messages.tsx [hunk 30]
```diff
@@ -0,0 +1,34 @@
+import * as React from 'react';
```
Comment (ghost) [inline frontend/src/components/messages/messages.tsx:1]: Remove all member names

File: frontend/src/components/messages/components/message/styles.module.scss [hunk 31]
```diff
@@ -0,0 +1,50 @@
+.message {
+  width: 358px;
+  display: flex;
+  flex-wrap: wrap;
+
+  align-self: flex-start;
+  flex-direction: row;
+
+  img {
+    position: relative;
```
Comment (ghost) [inline frontend/src/components/messages/components/message/styles.module.scss]:
```suggestion
```

File: frontend/src/components/messages/components/message-list/styles.module.scss [hunk 32]
```diff
@@ -0,0 +1,36 @@
...
+    display: flex;
+    flex-direction: column;
+  }
+
+  &::-webkit-scrollbar {
+    width: 5px;
+  }
+
+  &::-webkit-scrollbar-track {
+    background-color: #faf9f9; // var(--color-gray-light);
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/styles.module.scss]:
```suggestion
    background-color: #faf9f9;
```

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 33]
```diff
@@ -0,0 +1,51 @@
+import * as React from 'react';
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx:1]: Remove all member names

File: frontend/src/components/messages/components/member/member.tsx [hunk 34]
```diff
@@ -0,0 +1,32 @@
+import * as React from 'react';
```
Comment (ghost) [inline frontend/src/components/messages/components/member/member.tsx]: Remove all member names

File: frontend/src/components/messages/components/member-list/member-list.tsx [hunk 35]
```diff
@@ -0,0 +1,47 @@
+import * as React from 'react';
```
Comment (ghost) [inline frontend/src/components/messages/components/member-list/member-list.tsx]: Remove all member names

File: frontend/src/components/messages/components/input-message-form/styles.module.scss [hunk 36]
```diff
@@ -0,0 +1,20 @@
+.inputMessageForm {
+  // width: 100%;
```
Comment (ghost) [inline frontend/src/components/messages/components/input-message-form/styles.module.scss]:
```suggestion
```

File: frontend/src/components/messages/components/input-message-form/input-message-form.tsx [hunk 37]
```diff
@@ -0,0 +1,80 @@
...
+    defaultValues: {
+      to: selectedMemberId,
+      text: '',
+    },
+    mode: 'onChange',
+  });
+
+  const dispatch = useDispatch();
+
+  const onSubmit = (formData: IMessagePayload) => {
```
Comment (ghost) [inline frontend/src/components/messages/components/input-message-form/input-message-form.tsx]: Check **A9** criteria

File: backend/src/data/repositories/messages-repository.ts [hunk 38]
```diff
@@ -0,0 +1,125 @@
+import { Sequelize, Op } from 'sequelize';
+import { MemberModel, MessageModel } from '../models';
+import { IMember, IMessage } from '~/common/interfaces';
+import { MessageKey, MemberKey, OrderKey } from '~/common/enums';
+
+class MessagesRepository {
+  // public getMembersAsChats(): Promise<IMember[] | null> {
```
Comment (ghost) [inline backend/src/data/repositories/messages-repository.ts]: Remove all commented code

File: backend/src/services/messages/messages.service.ts [hunk 39]
```diff
@@ -0,0 +1,22 @@
+import { messagesRepository } from '~/data/repositories';
+import { IMember, IMessage } from '~/common/interfaces';
+
+class Messages {
+  // public getMembersAsChats(): Promise<IMember[] | null> {
+  //   return messagesRepository.getMembersAsChats();  // userId
+  // }
```
Comment (ghost) [inline backend/src/services/messages/messages.service.ts]:
```suggestion
```

File: backend/src/data/models/member.ts [hunk 40]
```diff
@@ -0,0 +1,29 @@
+import { Sequelize, DataTypes, ModelCtor, Model } from 'sequelize';
```
Comment (ghost) [inline backend/src/data/models/member.ts]: Remove it

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/store/slices/messages/messages.slice.ts (hunk 26 above)
Comment (petrynchyn) [reply frontend/src/store/slices/messages/messages.slice.ts]:
```
/messages.slice.ts(87,40):
Property 'id' does not exist on type 'IMember[]'.  TS2339

    85 |     dispatch(actions.setMembers(response));
    86 |     response?.length
  > 87 |       ? dispatch(selectMember(response.id))
       |                                        ^
    88 |       : dispatch(actions.setMessages([]));
    89 |
    90 |   } catch (error) {
```

Comment (ghost) [reply frontend/src/store/slices/messages/messages.slice.ts]: Make it works 🙂
We don't want to get an array

Comment (petrynchyn) [reply frontend/src/store/slices/messages/messages.slice.ts]: something like that?
```
const firstUserId: string | undefined = response.find(user => user !== undefined)?.id;
firstUserId
  ? dispatch(selectMember(firstUserId))
  : dispatch(actions.setMessages([]));
```

File: backend/src/data/repositories/index.ts [hunk 41]
```diff
@@ -17,6 +18,7 @@ const doctorRepository = new DoctorRepository();
 const appointment = new Appointment();
 const notification = new Notification();
 const permissionRepository = new PermissionRepository();
+const messagesRepository = new MessagesRepository();
```
Comment (ghost) [inline backend/src/data/repositories/index.ts]:
```suggestion
const permission = new Permission();
const message = new Message();
```

File: backend/src/data/repositories/index.ts [hunk 42]
```diff
@@ -7,6 +7,7 @@ import { Document } from './document-repository';
 import { Notification } from './notification.repository';
 import { DoctorRepository } from './doctor-repository';
 import { PermissionRepository } from './permission-repository';
+import { MessagesRepository } from './messages-repository';
```
Comment (ghost) [inline backend/src/data/repositories/index.ts]:
```suggestion
import { MessagesRepository } from './messages.repository';
```
Rename all by this way

File: backend/src/data/repositories/messages-repository.ts [hunk 43]
```diff
@@ -0,0 +1,44 @@
+import { Op } from 'sequelize';
+import { MessageModel } from '../models';
+import { IMessage } from '~/common/interfaces';
+import { MessageKey, SortType } from '~/common/enums';
+
+class MessagesRepository {
```
Comment (ghost) [inline backend/src/data/repositories/messages-repository.ts]:
```suggestion
class Message {
```

File: backend/src/data/repositories/messages-repository.ts [hunk 44]
```diff
@@ -0,0 +1,44 @@
+import { Op } from 'sequelize';
```
Comment (ghost) [inline backend/src/data/repositories/messages-repository.ts]: Rename to `backend/src/data/repositories/message.repository.ts`

File: backend/src/data/repositories/messages-repository.ts [hunk 45]
```diff
@@ -0,0 +1,44 @@
...
+      },
+
+      order: [
+        [MessageKey.CREATED_AT, SortType.DESC],
+      ],
+
+    });
+  }
+
+  public createMessage(message: IMessage): Promise<IMessage | null> {
```
Comment (ghost) [inline backend/src/data/repositories/messages-repository.ts]:
```suggestion
  public createMessage(message: IMessage): Promise<IMessage> {
```

File: backend/src/data/repositories/messages-repository.ts [hunk 46]
```diff
@@ -0,0 +1,44 @@
+import { Op } from 'sequelize';
+import { MessageModel } from '../models';
+import { IMessage } from '~/common/interfaces';
+import { MessageKey, SortType } from '~/common/enums';
+
+class MessagesRepository {
+
+  public getMessagesByUserId(toUserId: string, userId: string): Promise<IMessage[] | null> {
```
Comment (ghost) [inline backend/src/data/repositories/messages-repository.ts]:
```suggestion
  public getMessagesByUserId(toUserId: string, userId: string): Promise<IMessage[]> {
```

File: backend/src/data/repositories/user-repository.ts [hunk 47]
```diff
@@ -126,6 +127,28 @@ class User {
     });
   }
 
+  public filterUsersByName(name: string): Promise<IUser[] | null> {
```
Comment (ghost) [inline backend/src/data/repositories/user-repository.ts]:
```suggestion
  public filterUsersByName(name: string): Promise<IUser[]> {
```

File: backend/src/services/messages/messages.service.ts [hunk 48]
```diff
@@ -0,0 +1,15 @@
+import { messagesRepository } from '~/data/repositories';
+import { IMessage } from '~/common/interfaces';
+
+class Messages {
```
Comment (ghost) [inline backend/src/services/messages/messages.service.ts]:
```suggestion
class Message {
```

File: backend/src/services/messages/messages.service.ts [hunk 49]
```diff
@@ -0,0 +1,15 @@
+import { messagesRepository } from '~/data/repositories';
```
Comment (ghost) [inline backend/src/services/messages/messages.service.ts]: Rename to `backend/src/services/messages/message.service.ts`

File: backend/src/services/user-service/user-service.service.ts [hunk 50]
```diff
@@ -23,6 +23,10 @@ class User {
     return userRepository.getDoctorDetailsById(id);
   }
 
+  public filterUsersByName(name: string): Promise<IUser[] | null> {
```
Comment (ghost) [inline backend/src/services/user-service/user-service.service.ts]:
```suggestion
  public filterUsersByName(name: string): Promise<IUser[]> {
```

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 51]
```diff
@@ -0,0 +1,70 @@
...
+  InputType,
+  InputColor,
+  ButtonType,
+  ButtonColor,
+  ButtonStyleType,
+  UserKey,
+} from 'common/enums';
+import { TextInput, Button } from 'components/common';
+import { MessagesActionCreator } from 'store/slices';
+import { debounce } from 'common/helpers';
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]:
```suggestion
import { debounce } from 'helpers';
```
move helper from `common/helpers` to `helpers`

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 52]
```diff
@@ -0,0 +1,70 @@
...
+    defaultValues: {
+      [UserKey.NAME]: null,
+    },
+    mode: 'onChange',
+  });
+
+  const dispatch = useDispatch();
+
+  const handleFormSubmit = React.useCallback(() => {
+    const name = getValues()[UserKey.NAME] || '*';
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]:
```suggestion
    const name = getValues()[UserKey.NAME] ?? DEAULT_SEARCH_VALUE;
```

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 53]
```diff
@@ -0,0 +1,70 @@
...
+  className?: string;
+}
+
+const AddUserForm: React.FC<Props> = ({ className }) => {
+
+  const { handleSubmit, errors, control, formState: { isDirty }, getValues } = useForm({
+    defaultValues: {
+      [UserKey.NAME]: null,
+    },
+    mode: 'onChange',
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]: Why change if you use it as submit? 🤔

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 54]
```diff
@@ -0,0 +1,70 @@
...
+      <div className={styles.submitBtn}>
+        <Button
+          label="New chat"
+          hasHiddenLabel={false}
+          type={ButtonType.SUBMIT}
+          color={ButtonColor.PRIMARY_DARK}
+          styleType={ButtonStyleType.WITHOUT_BORDER}
+          isDisabled={!isDirty}
+        />
+      </div>
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]:
```suggestion
    <form onChange={handleChange} className={clsx(styles.addUserForm, className)}>
```
Don't forget to remove styles

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 55]
```diff
@@ -0,0 +1,70 @@
...
+      <div className={styles.inputText}>
+        <TextInput
+          name={UserKey.NAME}
+          placeholder="Type a name..."
+          type={InputType.TEXT}
+          label=""
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]: Always add labels

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 56]
```diff
@@ -0,0 +1,51 @@
...
+
+import styles from './styles.module.scss';
+
+interface Props {
+  className?: string;
+}
+
+const dateOptions = { day: 'numeric', month: 'short' };
+const timeOptions = { hour: '2-digit', minute: '2-digit', hour12:false };
+const locale = 'en-GB';
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx]:
```suggestion

```

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 57]
```diff
@@ -0,0 +1,51 @@
...
+    messages,
+    selectedUserAvatarPath: selectedUser?.imagePath,
+  }));
+
+  return (
+    <div className={clsx(styles.messageList, className)}>
+      {messages.map(({ id, userId: messageUserId, text, createdAt }, index, array) => {
+
+        const curFromNow = new Date(createdAt).toLocaleDateString(locale, dateOptions);
+        const prevFromNow = new Date(array[index + 1]?.createdAt).toLocaleDateString(locale, dateOptions);
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx]:
```suggestion
```

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 58]
```diff
@@ -0,0 +1,51 @@
...
+
+        return (
+          <div key={id} className={styles.messageContainer}>
+
+            {(curFromNow !== prevFromNow || (array.length + 1) === index)
+              && <HorizontalLine label={curFromNow} />}
+
+            <Message
+              message={text}
+              time={new Date(createdAt).toLocaleTimeString(locale, timeOptions)}
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx]: Use `getFormattedDate` helper

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 59]
```diff
@@ -0,0 +1,51 @@
...
+    <div className={clsx(styles.messageList, className)}>
+      {messages.map(({ id, userId: messageUserId, text, createdAt }, index, array) => {
+
+        const curFromNow = new Date(createdAt).toLocaleDateString(locale, dateOptions);
+        const prevFromNow = new Date(array[index + 1]?.createdAt).toLocaleDateString(locale, dateOptions);
+
+        return (
+          <div key={id} className={styles.messageContainer}>
+
+            {(curFromNow !== prevFromNow || (array.length + 1) === index)
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx]: Add constant for magic numbers

File: frontend/src/components/messages/components/message-list/styles.module.scss [hunk 60]
```diff
@@ -0,0 +1,36 @@
+.messageList {
+  display: flex;
+  flex-direction: column-reverse;
+
+  overflow-y: auto;
+  height: 100%;
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/styles.module.scss]: It don't work

File: frontend/src/services/messages/messages.service.ts [hunk 61]
```diff
@@ -0,0 +1,41 @@
...
+class Messages {
+  #http: Http;
+  #apiPrefix: string;
+
+  constructor({ http, apiPrefix }: Constructor) {
+    this.#http = http;
+    this.#apiPrefix = apiPrefix;
+  }
+
+  public loadUserMessages(toUserId: string | undefined): Promise<IMessage[]> {
```
Comment (ghost) [inline frontend/src/services/messages/messages.service.ts]:
```suggestion
  public loadUserMessages(toUserId: string): Promise<IMessage[]> {
```

File: frontend/src/services/user-api/user-api.service.ts [hunk 62]
```diff
@@ -72,5 +72,14 @@ class UserApi {
       method: HttpMethod.GET,
     });
   }
+
+  public filterUsersByName(name: string | undefined): Promise<IUser[]> {
```
Comment (ghost) [inline frontend/src/services/user-api/user-api.service.ts]:
```suggestion
  public filterUsersByName(name: string): Promise<IUser[]> {
```

File: frontend/src/store/slices/messages/messages.slice.ts [hunk 63]
```diff
@@ -0,0 +1,99 @@
...
+
+const loadFilteredUsersAsChats = (name: string): AppThunk => async dispatch => {
+  try {
+    const response = await userApi.filterUsersByName(name);
+    dispatch(actions.setUsers(response));
+
+    const firstUserId: string | undefined = response.find(user => user !== undefined)?.id;
+    firstUserId
+      ? dispatch(selectUser(firstUserId))
+      : dispatch(actions.setMessages([]));
```
Comment (ghost) [inline frontend/src/store/slices/messages/messages.slice.ts]:
```suggestion
   
      dispatch(firstUserId ? selectUser(firstUserId) : actions.setMessages([]))
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: backend/src/services/messages/message.service.ts [hunk 64]
```diff
@@ -0,0 +1,15 @@
+import { messages as messagesRepository } from '~/data/repositories';
+import { IMessage } from '~/common/interfaces';
+
+class Message {
+
+  public getMessagesByUserId(toUserId: string, userId: string): Promise<IMessage[] | null> {
```
Comment (ghost) [inline backend/src/services/messages/message.service.ts]:
```suggestion
  public getMessagesByUserId(toUserId: string, userId: string): Promise<IMessage[]> {
```

File: backend/src/services/messages/message.service.ts [hunk 65]
```diff
@@ -0,0 +1,15 @@
...
+    return messagesRepository.getMessagesByUserId(toUserId, userId);
+  }
+
+  public createMessage(message: IMessage): Promise<IMessage | null> {
```
Comment (ghost) [inline backend/src/services/messages/message.service.ts]:
```suggestion
  public createMessage(message: IMessage): Promise<IMessage> {
```

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 66]
```diff
@@ -0,0 +1,56 @@
...
+
+interface Props {
+  className?: string;
+}
+
+const AddUserForm: React.FC<Props> = ({ className }) => {
+
+  const { errors, control, getValues } = useForm({
+    defaultValues: {
+      [UserKey.NAME]: null,
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]:
```suggestion
      [UserKey.NAME]: '',
```

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 67]
```diff
@@ -0,0 +1,56 @@
...
+    },
+  });
+
+  const dispatch = useDispatch();
+
+  const handleFormSubmit = React.useCallback(() => {
+    const name = getValues()[UserKey.NAME] || DEAULT_SEARCH_VALUE;
```
Comment (ZubkoWhatislove) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx:32]:
```suggestion
    const name = getValues()[UserKey.NAME] ?? DEAULT_SEARCH_VALUE;
```

File: frontend/src/components/messages/components/add-user-form/add-user-form.tsx [hunk 68]
```diff
@@ -0,0 +1,56 @@
...
+import clsx from 'clsx';
+import {
+  InputType,
+  InputColor,
+  UserKey,
+} from 'common/enums';
+import { TextInput } from 'components/common';
+import { MessagesActionCreator } from 'store/slices';
+import { debounce } from 'helpers';
+import { DELAY_TIMEOUT, DEAULT_SEARCH_VALUE } from './constants';
```
Comment (ghost) [inline frontend/src/components/messages/components/add-user-form/add-user-form.tsx]:
```suggestion
import { DELAY_TIMEOUT, DEAULT_SEARCH_VALUE } from './common/constants';
```

File: frontend/src/components/messages/components/message-list/message-list.tsx [hunk 69]
```diff
@@ -0,0 +1,51 @@
...
+import { getFormattedDate } from 'helpers';
+import { DateFormat } from 'common/enums';
+
+import styles from './styles.module.scss';
+
+interface Props {
+  className?: string;
+}
+
+const deltaIndex = 1;
```
Comment (ghost) [inline frontend/src/components/messages/components/message-list/message-list.tsx]: should be constant

File: frontend/src/store/slices/messages/messages.slice.ts [hunk 70]
```diff
@@ -0,0 +1,99 @@
...
+    throw error;
+  }
+};
+
+const loadFilteredUsersAsChats = (name: string): AppThunk => async dispatch => {
+  try {
+    const response = await userApi.filterUsersByName(name);
+    dispatch(actions.setUsers(response));
+
+    const firstUserId: string | undefined = response.find(user => user !== undefined)?.id;
```
Comment (ghost) [inline frontend/src/store/slices/messages/messages.slice.ts]: Only user should select the user

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — DISMISSED** [review]
👍

**Review by ghost — APPROVED** [review]
👍

**Pull request merged**